src/Java.Interop.Tools.JavaTypeSystem/JavaModels/JavaClassModel.cs

```diff
@@ -0,0 +1,118 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Java.Interop.Tools.JavaTypeSystem.Models
+{
+	public class JavaClassModel : JavaTypeModel
+	{
+		IDictionary<JavaTypeReference, JavaTypeReference>? generic_inheritance_mapping;
+
+		public string BaseType { get; }
+		public string BaseTypeGeneric { get; }
+		public string BaseTypeJni { get; }
+
+		public JavaTypeReference? BaseTypeReference { get; private set; }
+		public List<JavaConstructorModel> Constructors { get; } = new List<JavaConstructorModel> ();
+
+		public JavaClassModel (JavaPackage javaPackage, string javaNestedName, string javaVisibility, bool javaAbstract, bool javaFinal, string javaBaseType, string javaBaseTypeGeneric, string javaDeprecated, bool javaStatic, string jniSignature, string baseTypeJni) :
+			base (javaPackage, javaNestedName, javaVisibility, javaAbstract, javaFinal, javaDeprecated, javaStatic, jniSignature)
+		{
+			BaseType = javaBaseType;
+			BaseTypeGeneric = javaBaseTypeGeneric;
+			BaseTypeJni = baseTypeJni;
+		}
+
+		public override void Resolve (JavaTypeCollection types, List<JavaUnresolvableModel> unresolvables)
+		{
+			var type_parameters = GetApplicableTypeParameters ().ToArray ();
+
+			// Resolve base class
+			if (FullName != "java.lang.Object" && FullName != "java.lang.Throwable") {
+				try {
+					BaseTypeReference = types.ResolveTypeReference (TypeResolutionOptions.ResolveGenerics ? BaseTypeGeneric : BaseType, type_parameters);
```

Comment:
Yet another reason why "overloading" `Resolve` is "weird": `ResolveTypeReference()` returns a value, not `void`.